lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -363,211 +642,256 @@ runSelectionStep lens s
 
+-- | Constructs change bundles for a set of selected inputs and outputs.
+--
+-- Returns 'Nothing' if the specified inputs do not provide enough ada to
+-- satisfy the minimum delta and minimum ada quantities of the change bundles
+-- generated.
+--
+-- This function will generate runtime errors if:
+--
+--    1.  The total balance of all outputs is not less than or equal to the
+--        total balance of all inputs.
+--
+--    2.  The total ada balance of all outputs is zero.
+--
+-- Pre-condition (1) should be satisfied by any result produced by the
+-- 'runSelection' function.
+--
+-- Pre-condition (2) should be satisfied by assigning a minimum ada quantity
+-- to every output token bundle.
+--
 makeChange
-    :: NonEmpty TokenBundle
-        -- ^ Token bundles of selected inputs
+    :: (TokenMap -> Coin)
+        -- A function that computes the minimum required ada quantity for a
+        -- particular output.
+    -> Coin
+        -- ^ The minimal (and optimal) delta between the total ada balance
+        -- of all input bundles and the total ada balance of all output and
+        -- change bundles, where:
+        --
+        --    delta = getCoin (fold inputBundles)
+        --          - getCoin (fold outputBundles)
+        --          - getCoin (fold changeBundles)
+        --
+        -- This typically captures fees plus key deposits.
+        --
+    -> Maybe Coin
+        -- ^ An optional extra source of ada.
     -> NonEmpty TokenBundle
-        -- ^ Token bundles of original outputs
+        -- ^ Token bundles of selected inputs.
     -> NonEmpty TokenBundle
-        -- ^ Change bundles
-makeChange inputBundles outputBundles
+        -- ^ Token bundles of original outputs.
+    -> Either UnableToConstructChangeError (NonEmpty TokenBundle)
+        -- ^ Generated change bundles.
+makeChange
+    minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBundles
     | not (totalOutputValue `leq` totalInputValue) =
         totalInputValueInsufficient
-    | totalOutputCoinValue == Coin 0 =
+    | TokenBundle.getCoin totalOutputValue == Coin 0 =
         totalOutputCoinValueIsZero
-    | otherwise =
-        change
+    | otherwise = do
+            -- The following subtraction is safe, as we have already checked
+            -- that the total input value is greater than the total output
+            -- value:
+        let excess :: TokenBundle
+            excess =
+                totalInputValue `TokenBundle.unsafeSubtract` totalOutputValue
+
+        let (excessCoin, excessAssets) = TokenBundle.toFlatList excess
+
+        let unknownAssets =
+                Map.toList $ F.foldr discardKnownAssets mempty inputBundles
+
+        let changeForKnownAssets :: NonEmpty TokenMap
+            changeForKnownAssets = F.foldr
+                (NE.zipWith (<>) . makeChangeForKnownAsset outputTokens)
+                (TokenMap.empty <$ outputTokens)
+                excessAssets
```

Comment:
It is doing a little extra work indeed. But partitioning the list upfront isn't necessarily less work. Here, we got the change for known and unknown assets in two list traversals. Though, when partitioning before, you end up with an extra first traversals for creating the partition (although, the next two are done on smaller lists)

In the end, I privileged simplicity over optimizations, because I find this last version easier to read. It took me some time to understand what exactly the previous version was doing, but more importantly, why it was still correct. It required more reasoning. Here, it's quite straightforward IMO.

Note: I've built the source with profiling yesterday evening and intended to run some profiling analysis on this entire code today. I've tried running `performSelection` on very large UTxO (> 50k) and it didn't went well. If this appears to be one of the bottleneck, I am happy to rework the implementation in favor of optimizations.

Reply:
> I've tried running performSelection on very large UTxO (> 50k) and it didn't went well.

Just out of curiosity, how were you generating your very large UTxO sets? Were they taken from an existing wallet DB, or did you use a QC generator?

Reply:
I used the QC generator. So arguably, the generator was already taking a lot of the time.

Reply:
> I used the QC generator. So arguably, the generator was already taking a lot of the time.

That would also be my assumption.

I guess one way to check is to always start from the same very large UTxO set, and run multiple selections on that single set. Each selection should still be different, as the selection is still random (provided that we start with a different seed).